pytradfri/resource.py

```diff
@@ -5,39 +5,102 @@
+    ROOT_SIGNAL_REPEATER,
+)
+from .device.air_purifier import AirPurifierResponse
+from .device.blind import BlindResponse
+from .device.light import LightResponse
+from .device.socket import SocketResponse
 
 # type alias
 TypeRaw = Dict[str, Union[str, int, List[Dict[str, Union[str, int]]]]]
 TypeRawSimple = Dict[str, Union[str, int]]
 TypeRawList = Dict[str, List[Dict[str, Union[str, int]]]]
 
 
+class DeviceInfoResponse(BaseModel):
+    """Represent the device info part of the device response."""
+
+    manufacturer: str = Field(alias=ATTR_DEVICE_MANUFACTURER)
+    model_number: str = Field(alias=ATTR_DEVICE_MODEL_NUMBER)
+    serial: str = Field(alias=ATTR_DEVICE_SERIAL)
+    firmware_version: str = Field(alias=ATTR_DEVICE_FIRMWARE_VERSION)
+    power_source: int | None = Field(alias=ATTR_DEVICE_POWER_SOURCE)
+    battery_level: int | None = Field(alias=ATTR_DEVICE_BATTERY)
+
+
+class ApiResourceResponse(BaseModel):
+    """Represent an API response."""
+
+    application_type: int | None = Field(alias=ATTR_APPLICATION_TYPE)
+    created_at: int = Field(alias=ATTR_CREATED_AT)
+    device_info: DeviceInfoResponse | None = Field(alias=ATTR_DEVICE_INFO)
+    id: int = Field(alias=ATTR_ID)
+    last_seen: int | None = Field(alias=ATTR_LAST_SEEN)
+    name: str | None = Field(alias=ATTR_NAME)
+    ota_update_state: int | None = Field(alias=ATTR_OTA_UPDATE_STATE)
+    reachable: int | None = Field(alias=ATTR_REACHABLE_STATE)
+
+    air_purifier: list[AirPurifierResponse] | None = Field(alias=ROOT_AIR_PURIFIER)
+    blind: list[BlindResponse] | None = Field(alias=ATTR_START_BLINDS)
+    light: list[LightResponse] | None = Field(alias=ATTR_LIGHT_CONTROL)
+    socket: list[SocketResponse] | None = Field(alias=ATTR_SWITCH_PLUG)
+    signal_repater: list[SignalRepeaterResponse] | None = Field(
+        alias=ROOT_SIGNAL_REPEATER
+    )
+
+
 class ApiResource:
     """Base object for resources returned from the gateway."""
 
-    def __init__(self, raw: ApiResourceResponse) -> None:
+    _model_class: type[ApiResourceResponse] = ApiResourceResponse
+    raw: ApiResourceResponse
+
+    def __init__(self, raw: TypeRawList) -> None:
         """Initialize base object."""
-        self.raw = raw
+        self.raw = self._model_class(**raw)
 
     @property
-    def id(self) -> str | None:
+    def id(self) -> int | None:
         """Id."""
-        return self.raw[ATTR_ID]
+        return self.raw.id
 
     @property
     def name(self) -> str | None:
         """Name."""
-        return self.raw[ATTR_NAME]
+        return self.raw.name
 
     @property
     def created_at(self) -> datetime | None:
         """Return timestamp of creation."""
-        if ATTR_CREATED_AT not in self.raw:
-            return None
-        return datetime.utcfromtimestamp(self.raw[ATTR_CREATED_AT])
+        if self.raw.created_at:
+            return datetime.utcfromtimestamp(self.raw.created_at)
+
+        return None
 
     @property
     @abstractmethod
```

Comment:
Update the `observe_callback` and `process_result` below in a similar way to keep compatibility with the other resources that we don't update now.